[WIP] Multiple Codec Support and General cleanup

mcdonnelldean opened the pull request:
Ok @mcollina,

Unfortunately this one got a bit messy because I was trying to clean up and implement multi codec support. However, I think the result is good in the end, it's broken down a bit more, names are cleaned up too. Still need some tests but provisionally, the following are done, #6 and #7 also #13 and #14 should now be easier.

mcdonnelldean committed:
design work for adding multiple codecs

mcdonnelldean commented on hyperemitter.js:
```diff
@@ -58,8 +58,8 @@ function HyperEmitter (db, schema, opts) {
   })
 
   function process (change, enc, next) {
```
I removed the call to headers directly. We don't expose .messages (.codecs) so there was no reason to use one over the other. At this point they are merged anyway.

mcollina commented on hyperemitter.js:
```diff
@@ -26,7 +26,7 @@ function HyperEmitter (db, schema, opts) {
   }
 
-  this.messages = protobuf(schema)
+  this.codecs = protobuf(schema)
```
let's just stop accepting a schema and wrapping protobuf, we will just support codecs, and otherwise we emit buffers.

mcdonnelldean replied:
I think this makes the most sense. On it now

mcollina replied:
On second thought, I don't want to bump the major now. So, if the argument
is a string, keep the old behavior, but add a deprecation warnig. And add
an alias from 'codecs' to messages. You can use an ES5 object property for
that, and add a deprecation warning too.
Il giorno dom 24 mag 2015 alle 12:50 Dean McDonnell <
notifications@github.com> ha scritto:

> In hyperemitter.js
> https://github.com/mcollina/hyperemitter/pull/15#discussion_r30951723:
> 
> > @@ -26,7 +26,7 @@ function HyperEmitter (db, schema, opts) {
> >      return new HyperEmitter(db, schema, opts)
> >    }
> > -  this.messages = protobuf(schema)
> > -  this.codecs = protobuf(schema)
> 
> I think this makes the most sense. On it now
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/mcollina/hyperemitter/pull/15/files#r30951723.

mcdonnelldean replied:
Cool, I don't understand the first part though. If which argument is a string? I presume you mean ensure we can add the protobuf schema so we don't break existing peeps? The new sig is slightly different. I'd probably lease schema there then if you are worried about breaking things?

mcollina replied:
Yeah, leave schema there, but make it fully optional.
Il giorno dom 24 mag 2015 alle 13:09 Dean McDonnell <
notifications@github.com> ha scritto:

> In hyperemitter.js
> https://github.com/mcollina/hyperemitter/pull/15#discussion_r30951803:
> 
> > @@ -26,7 +26,7 @@ function HyperEmitter (db, schema, opts) {
> >      return new HyperEmitter(db, schema, opts)
> >    }
> > -  this.messages = protobuf(schema)
> > -  this.codecs = protobuf(schema)
> 
> Cool, I don't understand the first part though. If which argument is a
> string? I presume you mean ensure we can add the protobuf schema so we
> don't break existing peeps? The new sig is slightly different. I'd probably
> lease schema there then if you are worried about breaking things?
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/mcollina/hyperemitter/pull/15/files#r30951803.

mcdonnelldean replied:
:thumb:

mcdonnelldean committed:
see description
- removed invalid test
- added back in messages
- depreciated buffers
- changed emit and process to handle changes

mcdonnelldean commented:
@mcollina Just wanted to check if I'm on the right track here?

mcollina commented on hyperemitter.js:
```diff
@@ -21,12 +21,20 @@ var defaults = {
+  this.status = new EventEmitter()
+
+  if (Buffer.isBuffer(this.codecs)) {
+    this.status.emit('buffers are being depreciated')
```
Don't emit, just use `console.log`, or better https://nodejs.org/api/util.html#util_util_deprecate_function_string.

You should also check for strings, not only Buffers.

Also, there is a spelling mistake :P.

mcdonnelldean replied:
@mcollina What should I do with a string? Is that an acceptable input to Protobuf?

mcollina replied:
Yes it is.

mcollina commented on hyperemitter.js:
```diff
@@ -106,21 +117,17 @@ function HyperEmitter (db, schema, opts) {
-    else throw err
-  }
+  if (encoder) data = encoder.encode(data)
+  else this.status.emit('no encoder found for ' + name)
```
don't emit something like this.

I think that:
1. if the argument is a String or a Buffer, then we can send them through as Buffers, and emit them as such.
2. if it's an object, then we should return an error, just as before.

mcdonnelldean replied:
Noted

mcollina commented on hyperemitter.js:
```diff
@@ -140,6 +147,20 @@ HyperEmitter.prototype.on = function (name, callback) {
+    name.forEach(function (set) {
+      that.codecs[set.name] = set.codec
+    })
+  }
```
missing `return this` at the end. I want a chainable API.

mcdonnelldean replied:
noted

mcollina commented on hyperemitter.js:
```diff
@@ -58,13 +66,16 @@ function HyperEmitter (db, schema, opts) {
-  // the status of this HyperEmitter
-  this.status = new EventEmitter()
+    if (decoder) event = decoder.decode(event)
+    else that.status.emit('no decoder found for ' + name)
```
Instead of emitting a status message, just emit the event name with the Buffer as an argument.

mcdonnelldean replied:
noted.

mcollina commented:
You are on the right track. Be sure to update tests along the way, as I think there are some functionality changes.

mcdonnelldean commented:
@mcollina Awesome, yeah this was first pass, I have a few tests to write to complete. On potential issues or maybe just a question.

If we have an codec we will encode / decode. If we don't have it we wont. There is a potential issue in that the payload may be encoded on one side but missing the decoder on the other side. Should the event container specify if it's payload is encoded or not?

Also, should we have a flag that allows the user to turn on missing encoder / decoder messages. I feel that a lot of the status.emit messages would be better as log messages.

mcollina commented:
I think we should have a `unknownEvent` handler in the options, with a default.

> If we have an codec we will encode / decode. If we don't have it we wont. There is a potential issue in that the payload may be encoded on one side but missing the decoder on the other side. Should the event container specify if it's payload is encoded or not?

I don't think so. If that happens it's either a user-wanted thing or not, and we can control it via a construction flag. There are two cases:
1. the user is running two different codebases, so they possibly have different events.
2. the user is running two versions of the same codebase, and the new one has new events.

So, in case one the user wants the 'ignore' behavior (decode as Buffers), while in the second it might want the 'error' behavior, so that the process is killed.

Either 1 or 2 should be the default. I'm more inclined to 1 (do not error, but risk losing messages).

BTW, do not put 'phrases' in the events, but use a single word.

mcdonnelldean committed:
See Description
- split down logic for clarity
- performed rename where it made sense
- improved buffer input
- added registerCodec
- renamed headers.proto to codec.proto
- update existing tests where needed

mcdonnelldean commented:
@mcollina. Could you review whenever you get a chance. Cheers 😊

mcollina commented:
This is awesome! Can you please add a couple of tests for the registerCodec method, and document that one in the README?

mcdonnelldean commented:
Will do, ill ping you later when they are done

mcdonnelldean committed:
registerCodec update
- added tests
- fixed bugs
- added readme info

mcdonnelldean committed:
Fixed Sunday afternoon typos :D

mcdonnelldean commented:
@mcollina Done :dancer:

mcollina commented:
Whoaaa!!

mcollina commented:
Released as 1.2.0!
